### ISSUE TITLE
Development plan for v0.10.0

### ISSUE THREAD

**Issue by kamo-naoyuki** [espnet/espnet#1795]
Feel free to fill. 

## General

- [ ] More smart installation #1559
- [x] [ESPnet Model Zoo](https://github.com/espnet/espnet_model_zoo) ~~More smart eco-system #2168~~
    - The procedures from downloading to inference are too complicated for new.
    - I think the tar file should include the repository itself in that time.
    - Using zenodo? It seems better  for such usage.
- [x] Support pytorch1.5
- [x] Support pytorch1.6
- [ ] CI for Windows/MaxOSX 

## TTS

- [ ] VAE-based Tacotron2 (e.g. [GMVAE-Tacotron2](https://arxiv.org/pdf/1810.07217.pdf))
	- Tensorflow implementation (Thanks @rishikksh20)  
	  https://github.com/rishikksh20/gmvae_tacotron  
      https://github.com/rishikksh20/vae_tacotron2
- [ ] Add neural vocoder synthesis stage in the recipe
- [ ] [Dynamic convolution attention](https://arxiv.org/abs/1910.10288)
- [x] ~~Re-design frontend modules~~ -> ESPnet2
- [x] ~~Re-design embedding inputs~~ -> ESPnet2
- [ ] Add difficult sentences for the generation 

### Objective/Subjective evaluation #1665

- [ ] ASR token error rate evaluation (both or either of CTC and Attention)
- [ ] spectral objective measure (e.g. RMSE, CD)
- [ ] [deepSpeechDistances](https://github.com/mbinkowski/DeepSpeechDistances)
- [ ] subjective evaluation instruction

## ASR 
- [x] Lightweight and Dynamic Convolutions. #1599
- [ ] multiple encoders
- [ ] merge ASR-TTS #818 
  - [ ] extend it with Transformer
- [ ] check initialization effects

### Toward fast decoding #1384

- implement other scorers
    - [x] RNNLM
    - [ ] Coverage
    - [x] N-gram @qmpzzpmq ?
    - [ ] Multihead decoder
- [ ] unify to use api_v2 for all other tasks including (probably we don't need additional implementation for some of them below).
    - asr_mix
    - mt
    - st
    - etc.? 

### new recipe
- [ ] heroico #1500
- [x] ELRA-S0030_02 #1711 

## ASR Transducer
- [x] Documentation and Tutorial (related to #1682) #2444
- [ ] Librispeech recipe (related to #1704)
- [x] Support CUDA-Warp RNN-T (https://github.com/1ytic/warp-rnnt) #2444
- [x] Add customizable architecture  for transformer-based parts #2444
  - [x] Add TDNN-BN-ReLu blocks #2444
  - [x] Add CausalConv1d blocks #2444
  - [x] Support Conformer blocks #2444
- [x] Variant beam search algorithms #2444

## ESPnet2
- [x] Support ASR Transducer #2533
- [ ] Support DNN-HMM Hybrid ASR #1648
- [ ] Support MT
- [ ] Support ST
- [ ] Support ASR mix
- [x] Show model summary #1867
- [x] Native torch amp #1836
- [ ] pytorch xla
- [x] Super large dataset #1801
- [ ] Torch script

### TTS related
- [x] documentation (#2312)
- [x] demo notebook
- [x] Implement on-the-fly text preprocessing
- [x] Support teacher forcing decoding
- Create recipes
    - [x] jsut (jp)
    - [x] ljspeech (en)
	- [x] csmsc (zh)
    - [x] vctk (en, multi) 
	- [x] jvs (jp, finetune) #2465 
- [x] Transformer #2134
- [x] Fastspeech #2149
- [x] GST #2139
- [x] FastSpeech2 #2218
- [x] Support multi-speaker in the template recipe
- [x] ~~Multi-speaker Tacotron 2~~ -> GST Taco2
- [x] ~~Multi-speaker Transformer~~ -> GST Transformer
- [x] ~~Multi-speaker FastSpeech~~ -> GST FastSpeech
- [x] ~~Multi-speaker FastSpeech2~~ -> GST FastSpeech2
- [ ] ASR token error rate evaluation 

### ASR related

- [x] Transformer LM
- [x] Batch decoding
  - [ ]  RNN Decoder
  - [x] Transformer Decoder
  - [x] RNN LM
  - [x] Transformer LM
- [x] Specaug
- [x] Batch bin Sampler
- [ ] Integrate ngram scorer
- [ ] Dynamic/Lightweight Convolution
- [x] Conformer

### Status of ESPnet2 ASR Recipes 

- [x] aishell @Emrys365 
- [x] ami @ftshijt 
- [ ] aurora4 @sas91 
- [x] babel @ftshijt 
- [x] chime4 @sas91 
- [ ] chime5 @b-flo 
- [x] commonvoice @ftshijt 
- [x] csj @YosukeHiguchi 
- [x] dirha_wsj @ruizhilijhu
- [ ] fisher_callhome_spanish
- [ ] fisher_swbd @YosukeHiguchi 
- [ ] hkust @Emrys365 
- [x] how2 @b-flo
- [ ] hub4_spanish @ruizhilijhu
- [ ] iwslt18
- [ ] iwslt19
- [ ] jnas @YosukeHiguchi 
- [x] jsut @YosukeHiguchi 
- [ ] libri_trans
- [x] librispeech @simpleoier #1687
- [ ] must_c
- [x] reverb @sas91 
- [ ] ru_open_stt
- [ ] swbd @YosukeHiguchi #1566
- [ ] tedlium2 @simpleoier #1687
- [ ] tedlium3 @simpleoier #1687
- [x] timit @Emrys365 
- [x] vivos @b-flo
- [x] voxforge @kamo-naoyuki 
- [x] yesno @b-flo
- [x] wsj @kamo-naoyuki 
- [x] zeroth korean

**Comment by kamo-naoyuki** [comment]
I'd like to make sure the branching rule because our rule was collapsed during development of espnet2.

- If a new feature will be implemented => next version
- Bug fix, refactoring, or new recipe without any drastic changing => Merge into Master

The reason why we need to prepare a development branch is to avoid merging the bug/issue related to new features into master. Such issues from early development makes new users confused.

**Comment by kamo-naoyuki** [comment]
I prepared `develop` branch instead of `v.0.8.0` because it's meaningless to switch development branch.

**Comment by kan-bayashi** [comment]
I want to summarize the functions which are not exported to espnet2 from espnet.

**Comment by kamo-naoyuki** [comment]
What part do you mean? I roughly wrote the items in the above table.

**Comment by kan-bayashi** [comment]
Sorry, I missed. I will update them.

**Comment by kamo-naoyuki** [comment]
@kan-bayashi Would you remove v.0.8.0 branch and change develop branch to protected?

**Comment by kan-bayashi** [comment]
Sure!

**Comment by SeanNaren** [comment]
Any timeline on getting RNN-Transducers into ESPNET2? Anything I can do to help here?

**Comment by sw005320** [comment]
Sounds good. Actually @b-flo is working on some major changes (customizable architecture) of RNN-T and @b-flo and @kamo-naoyuki  are discussing implementing it to espnet2.
@b-flo and @kamo-naoyuki, maybe you can continue the discussion in the GitHub issue so that @SeanNaren and other people could help your development.

**Comment by kamo-naoyuki** [comment]
I roughly prepared the place for RNN-T here.

https://github.com/espnet/espnet/blob/90ea8c788075d4b8b6bd71250268da61309cf2a2/espnet2/asr/espnet_model.py#L114-L135

The implementation for training part is not a hard matter, but we may need new beam-search module for inference part? I don't understand in detail.

The main modules for ESPnet2 was  developed almost only by me and I'd like anyone to join to the development. My prior matters are roughly:

1. Implement some system to handle super large corpus #1801 
1. Port TTS recipes
1. Reorganize installation system
1. Support GAN-like task for future development
1. Port Transducer
1. Create DNN-HMM recipes
1. Port batch decoding
1. Support native amp and torch xla

**Comment by b-flo** [comment]
I need to finish some other works and I'll focus on porting all transducer models to espnet2. I can't really give a timeline but I would like to start next month.

About beam search module, I have some other search algorithms for transducer (breadth-first, N-step constrained, etc) I already implemented in espnet1 but I didn't do a PR yet. We need some discussions but it would enable batch decoding and could be incluced in current beam-search module with some minor changes.

@kamo-naoyuki given your feedback on the "customizable architecture" we talked about on slack, I can work on a draft for espnet2 for all models.
